### ISSUE TITLE
Lookups in inventory are eagerly evaluated in 2.1.1.0, greatly impacting performance

### ISSUE THREAD

**Issue by kadaan** [ansible/ansible#17024]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

Lookup Plugin

##### ANSIBLE VERSION

```
ansible 2.1.1.0
  config file = /Users/kadaan/Source/ansible_2.1.1.0_lookup_regression/ansible.cfg
  configured module search path = Default w/o overrides
```
##### SUMMARY

The timing of when lookups defined in an inventory are resolved has changed between 2.1.0.0 and 2.1.1.0.  In 2.1.0.0, the lookups are not resolved until they are used.  In 2.1.1.0, the lookups are eagerly resolved.
##### STEPS TO REPRODUCE
1. Clone the repository: https://github.com/kadaan/ansible_2.1.1.0_lookup_regression
2. Run: `./ansible_2.1.0.0.sh`
3. Run: `./ansible_2.1.1.0.sh`
##### EXPECTED RESULTS

Lookups are evaluated lazily.

```
Using /Users/kadaan/Source/ansible_2.1.1.0_lookup_regression/ansible.cfg as config file
Loaded callback default of type stdout, v2.0

PLAYBOOK: test.yml *************************************************************
1 plays in playbooks/test.yml

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
task path: /Users/kadaan/Source/ansible_2.1.1.0_lookup_regression/playbooks/test.yml:5
DEBUG: Lookup Initializing
DEBUG: Lookup - lookup1
ok: [localhost] => {
    "msg": []
}

TASK [debug] *******************************************************************
task path: /Users/kadaan/Source/ansible_2.1.1.0_lookup_regression/playbooks/test.yml:6
DEBUG: Lookup Initializing
DEBUG: Lookup - lookup2
ok: [localhost] => {
    "msg": []
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0

real    0m4.204s
user    0m1.394s
sys 0m1.239s
```
##### ACTUAL RESULTS

Lookups are evaluated eagerly

```
Using /Users/kadaan/Source/ansible_2.1.1.0_lookup_regression/ansible.cfg as config file
Loaded callback default of type stdout, v2.0

PLAYBOOK: test.yml *************************************************************
1 plays in playbooks/test.yml

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupQ
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupP
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupS
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupR
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupU
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupT
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupW
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupV
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupY
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupX
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupZ
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupA
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupC
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupB
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupD
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupG
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupF
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupI
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupH
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupK
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupJ
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupM
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupL
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupO
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupN
ok: [localhost]

TASK [debug] *******************************************************************
task path: /Users/kadaan/Source/ansible_2.1.1.0_lookup_regression/playbooks/test.yml:5
DEBUG: Lookup Initializing
DEBUG: Lookup - lookup1
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupQ
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupP
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupS
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupR
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupU
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupT
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupW
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupV
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupY
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupX
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupZ
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupA
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupC
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupD
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupG
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupF
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupI
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupH
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupK
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupJ
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupM
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupL
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupO
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupN
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupB
ok: [localhost] => {
    "msg": []
}

TASK [debug] *******************************************************************
task path: /Users/kadaan/Source/ansible_2.1.1.0_lookup_regression/playbooks/test.yml:6
DEBUG: Lookup Initializing
DEBUG: Lookup - lookup2
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupQ
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupP
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupS
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupR
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupU
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupT
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupW
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupV
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupY
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupX
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupZ
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupA
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupC
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupD
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupG
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupF
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupI
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupH
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupK
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupJ
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupM
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupL
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupO
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupN
DEBUG: Lookup Initializing
DEBUG: Lookup - lookupB
ok: [localhost] => {
    "msg": []
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0


real    1m19.533s
user    0m7.208s
sys 0m5.953s
```

**Comment by Daryes** [comment]
I'm having the same behavior with a task rebooting the servers after an update, like this : 

```
serial : 10%

 tasks:
  - include: roles/linux/common-reboot/tasks/main.yml
    when: reboot_required is defined and reboot_required.stat.exists == true 
```

Before, I had a single [skipping: ...] for each server, but with v2.1.1.0, each action in the reboot task is visible.
Also, with the serial restriction, the play stops after the first batch,

**Comment by brickyard** [comment]
https://github.com/ansible/ansible/commit/6b286ee0c88f3252a27ff94f39d794566c3f7d6f#diff-76ffc0551f8cf3d6255500316568e60bR412

```
self._connection.set_host_overrides(host=self._host, hostvars=variables.get('hostvars', {}).get(self._host.name, {}))
```

This line causes the change of behavior.  

Any comment from @nitzmahone on whether this was intended?

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will be included in the 2.2.0 and 2.1.2 releases.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
